sourcemap/accessor.go

```diff
@@ -0,0 +1,76 @@
+package sourcemap
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/go-sourcemap/sourcemap"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type Accessor interface {
+	Fetch(id Id) (*sourcemap.Consumer, error)
+	Remove(id Id)
+}
+
+type SmapAccessor struct {
+	es    elasticsearch
+	cache *cache
+}
+
+func NewSmapAccessor(config Config) (*SmapAccessor, error) {
+	logp.Debug("sourcemap", "NewSmapAccessor created at Time.now %v for index", time.Now().Unix(), config.Index)
+
+	es, err := NewElasticsearch(config.ElasticsearchConfig, config.Index)
+	if err != nil {
+		return nil, err
+	}
+
+	cache, err := newCache(config.CacheExpiration, config.CacheCleanupInterval)
+	if err != nil {
+		return nil, err
+	}
+	return &SmapAccessor{
+		es:    es,
+		cache: cache,
+	}, nil
+}
+
+func (s *SmapAccessor) Fetch(id Id) (*sourcemap.Consumer, error) {
+	if !id.Valid() {
+		return nil, Error{
+			Msg:  fmt.Sprintf("Sourcemap Key Error for %v", id.String()),
+			Kind: KeyError,
+		}
+	}
+	consumer, found := s.cache.fetch(id)
+	if consumer != nil {
+		// avoid fetching ES again when key was already queried
+		// but no Sourcemap was found for it.
+		return consumer, nil
+	}
+	if found {
+		return nil, errSmapNotAvailable(id)
+	}
+	consumer, err := s.es.fetch(id)
+	if err != nil {
+		return nil, err
+	}
+	s.cache.add(id, consumer)
```

Comment:
This can be negative cache, to work with the comment about avoiding ES fetch above?  Is there a concern about timing between indexing a new sourcemap upload and this being called?  Is a shorter TTL on the negative cache possible?

Reply:
It is on purpose to add nil values to the cache if they cannot be found. 

Reason for this is:
We do not offer a seperate option to enable/disable sourcemapping, neither do we offer to enable/disable it per service. Right now, as soon as `frontend.enabled=true` sourcemapping is enabled. So for every incoming `trancsaction` and `error` the processor checks whether there is a sourcemap available and tries to apply it. For customers/services that want to have frontend support but haven't uploaded sourcemaps, this would mean that we would fetch ES for every stacktraceframe of every record. 

It would be possible to set a different TTL for negative cache values, however I'd rather avoid having another config option here. Hardcoding and not providing a config option would also be weird, when the general TTL is configurable.

Regarding timing issue:
Depending on the TTL it could ofc be that a sourcemap is uploaded but the cache still holds the nil value. TTL default is 5 mins, so I wouldn't be too concerned about this right now. 
As this feature is experimental my main concern with caching was more directed to not influence the general APM server and ES performance too much with it.

Reply:
Sounds good.  An option to disable sourcemapping while keeping frontend enabled seems like a great idea to start.

Reply:
https://github.com/elastic/apm-server/issues/472